inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/annotation/DynamicAnnotatorWorkflowActionBarItemGroup.java

```diff
@@ -89,48 +102,115 @@ public String getStateClass()
+                for (SourceDocument doc : randomList) {
+                    if (getUsersWorkingOnTheDocument(doc) < traits
+                            .getDefaultNumberOfAnnotations()) {
+                        if (documentService.listAnnotatableDocuments(project, user).get(doc) == null
+                                || (documentService.listAnnotatableDocuments(project, user).get(doc)
+                                        .getState().equals(NEW))) {
+                            getAnnotationPage().getModelObject().setDocument(doc,
+                                    documentService.listSourceDocuments(project));
+                            Optional<AjaxRequestTarget> target = RequestCycle.get()
+                                    .find(AjaxRequestTarget.class);
+                            getAnnotationPage().actionLoadDocument(target.orElse(null));
+                            return;
+                        }
+                    }
+                }
+                //No documents left
+                noDocumentsLeft();
+                break;
+
+
+            case ("Curriculum annotation"):
+
+                // TODO logic for Curriculum annotation
+
+                //No documents left
+                noDocumentsLeft();
+                break;
+
+            default:
+                // Default, simply go through the list and return the first document
+                for (Map.Entry<SourceDocument, AnnotationDocument> entry : documentService
+                        .listAnnotatableDocuments(project, user).entrySet()) {
+                    // First check if too many users are already working on the document
+                    if (((getUsersWorkingOnTheDocument(entry.getKey())
+                            + 1) <= dynamicWorkloadExtension.readTraits(currentWorkload)
+                                    .getDefaultNumberOfAnnotations())) {
+                        // Now check if there either is no annotation document yet created or its
+                        // state is NEW
+                        if (entry.getValue() == null || entry.getValue().getState().equals(NEW)) {
+                            getAnnotationPage().getModelObject().setDocument(entry.getKey(),
+                                    documentService.listSourceDocuments(project));
+                            Optional<AjaxRequestTarget> target = RequestCycle.get()
+                                    .find(AjaxRequestTarget.class);
+                            getAnnotationPage().actionLoadDocument(target.orElse(null));
+                            return;
+                        }
+                    }
+                }
+                //No documents left
+                noDocumentsLeft();
+                break;
             }
         });
         finishDocumentDialog.show(aTarget);
     }
+
+    public int getUsersWorkingOnTheDocument(SourceDocument aDocument)
+    {
+        return (int) documentService
```

Comment:
Instead of fetching all stuff from the DB and then filtering on it, implement a new query method which does the filtering already on the SQL level. If necessary open a corresponding PR on WebAnno.